relayer/processor/path_processor_internal.go

```diff
@@ -766,38 +795,55 @@ func (pp *PathProcessor) assembleAndSendMessages(
 	// Each assembleMessage call below will make a query on the source chain, so these operations can run in parallel.
 	var wg sync.WaitGroup
 
-	for i, msg := range messages.packetMessages {
-		wg.Add(1)
-		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
-	}
-
+	// connection messages are highest priority
+	om.connMsgs = make([]connectionMessageToTrack, len(messages.connectionMessages))
 	for i, msg := range messages.connectionMessages {
 		wg.Add(1)
 		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
 	}
 
-	for i, msg := range messages.channelMessages {
-		wg.Add(1)
-		go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+	wg.Wait()
+
+	if len(om.msgs) == 1 {
+		om.chanMsgs = make([]channelMessageToTrack, len(messages.channelMessages))
+		// only assemble and send channel handshake messages if there are no conn handshake messages
+		// this prioritizes connection handshake messages, useful if a connection handshake needs to occur before a channel handshake
```

Comment:
Is this a possible case that could occur though? A channel is tightly coupled to a specific connection so I would think that a channel handshake msg should never be sent before it's associated connection has completed handshake and is already in the open state.

Reply:
This is more of an in general order of operations change. This was done when debugging a failure to restore ordered ICA channels. There is a lot of noise in packet failures when there is a large packet queue and an ICA channel has a timeout closure. It was resolved by prioritizing channel messages over packet messages (only send packet messages if there are no channel messages). As an extension, connection messages are prioritized over channel messages.

Reply:
Ahh okay that makes sense to me!